libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -110,22 +167,41 @@ class state_history_log {
       }
 
       if (block_num < _end_block)
-         truncate(block_num);
-      log.seek_end(0);
+         truncate(block_num); //truncate is expected to always leave file pointer at the end
+      else if (!prune_blocks)
+         log.seek_end(0);
+      else if (prune_blocks && _begin_block != _end_block)
+         log.seek_end(-sizeof(uint32_t));  //overwrite the trailing block count marker on this write
+
+      //if we're operating on a pruned block log and this is the first entry in the log, make note of the feature in the header
+      if(prune_blocks && _begin_block == _end_block)
+         header.magic = ship_magic(get_ship_version(header.magic), ship_feature_pruned_log);
+
       uint64_t pos = log.tellp();
       write_header(header);
       write_payload(log);
       uint64_t end = log.tellp();
-      EOS_ASSERT(end == pos + state_history_log_header_serial_size + header.payload_size, chain::plugin_exception,
```

Comment:
Unneeded afaict. `cfile` already checks the return value on all its writes that it wrote everything it expected and throws an exception otherwise.

If you go back in time you'll see that originally `log` was a `std::fstream` and the code looked like
https://github.com/EOSIO/eos/blob/c7f026710038f91f39c2a0d293d7102654c9576f/plugins/state_history_plugin/include/eosio/state_history_plugin/state_history_log.hpp#L101-L106
In this case if the `std::fstream::write()` fails no exception is thrown you'd need to go inspect the stream's state and stuff. So this EOS_ASSERT appears to be there to check that everything was written. When the code was modified from `std::fstream` to `fc::cfile` this could have been removed.

Reply:
I think the issue is that there could be a mismatch between the `payload_size` in the `header` and the actual bytes written out by the `write_payload` function. That is the responsibility of the caller to ensure they match, but this check would catch it early if they don't.

Reply:
yep fair enough :+1: